Add support for ON CONFLICT DO UPDATE SET with EXCLUDED table

mxm opened the pull request:
This adds support for insert queries containing ON CONFLICT DO UPDATE SET
expressions with EXCLUDED table references. Tests verify identical semantics as
the ON DUPLICATE KEY statement.

Examples:

```
CREATE TABLE t1 (id int primary key, other string);
INSERT INTO t1 (id, other) values (id, 'str');

-- Set other to 'backing up' on duplicate key 1
INSERT into t1 (id, other) VALUES (1, 'test')
  ON CONFLICT DO UPDATE SET other = 'backing up';

-- which is identical to
INSERT into t1 (id, other) VALUES (1, 'test')
  ON DUPLICATE KEY UPDATE other = 'backing up';

-- Overwrite other with 'test'
INSERT into t1 (id, other) VALUES (1, 'test')
  ON CONFLICT DO UPDATE SET other = excluded.other;

-- which is identical to
INSERT into t1 (id, other) VALUES (1, 'test')
  ON DUPLICATE KEY SET other = values(other);
```

matriv commented on CHANGES.txt:
```diff
@@ -60,6 +60,11 @@ Breaking Changes
 =======
 
+- Added support for ``INSERT INTO ... ON CONFLICT DO UPDATE SET col = val``
+  which is identical to ``INSERT INTO ... ON DUPLICATE KEY UPDATE col = val``.
```
Do we deprecate the old syntax? If that is the case, we need another bullet point stating that.

matriv commented on sql/src/main/java/io/crate/analyze/relations/ExcludedFieldProvider.java:
```diff
@@ -0,0 +1,62 @@
+ *
+ * Otherwise it just calls the wrapped field provider.
+ */
+public class ExcludedFieldProvider implements FieldProvider<Symbol> {
```
maybe have a unit test for this class?

mxm replied:
There are analyzer tests which check the behavior. Doesn't hurt to add a unit test as well.

matriv replied:
yeah not so important, just for completeness...

mxm replied:
No, thanks for mentioning it. I've added a test.

matriv commented on sql/src/test/java/io/crate/analyze/InsertFromValuesAnalyzerTest.java:
```diff
@@ -916,34 +918,65 @@ public void testInsertIntoTableWithNestedPartitionedByColumnAndNullValue() throw
+    @Test
+    public void testInsertFromValuesWithWithDuplicateKeyAndUseOfExcluded() {
+        expectedException.expect(IllegalArgumentException.class);
+        expectedException.expectMessage("Column reference \"excluded.name\" has too many parts");
```
Shouldn't we add another if clause like: https://github.com/crate/crate/pull/7087/files#diff-c55569a976310945831be2cae906ec15R104 and have a message ``Can't use EXCLUDED outside ON CONFLICT``? Would seem more consistent to me.

mxm replied:
In an earlier design, it was like that because the analysis of the two was in the same place. I'd prefer to keep it like it is because we have to add extra logic to detect that and it doesn't add much value. The error message for any other table than "excluded" should be the same in on duplicate key clauses. It is just a precaution here to check that the excluded field provider is not in use.

mxm commented on sql-parser/src/test/java/io/crate/sql/parser/TestStatementBuilder.java:
```diff
@@ -783,6 +783,16 @@ public void testInsertStmtBuilder() throws Exception {
+        try {
+            printStatement("insert into t (a, b) values (1, 2) on conflict do nothing");
+        } catch (UnsupportedOperationException e) {
+            // this is what we want
```
thx fixed

mxm commented:
@mfussenegger @matriv @chaudum Thanks for your review. I've addressed your comments.

@matriv Considering the expectedException, I think it is ok to use try/catch/fail here because otherwise we have to duplicate a lot of code in extra test methods.
